File: 1769-minimum-number-of-operations-to-move-all-balls-to-each-box/1769-minimum-number-of-operations-to-move-all-balls-to-each-box.cpp
class Solution {
public:
    vector<int> minOperations(string boxes) {
        int n = boxes.size(), j;
        int ballsToLeft = 0, movesToLeft = 0;
        int ballsToRight = 0, movesToRight = 0;
        vector<int> ans(n);

        // Single pass: Calculate both left and right
        for (int i = 0; i < n; ++i){
            // Left pass
            ans[i] += movesToLeft;
            ballsToLeft += boxes[i] - 48;
            movesToLeft += ballsToLeft;

            // Right pass
            j = n - 1 - i;
            ans[j] += movesToRight;
            ballsToRight += boxes[j] - 48;
            movesToRight += ballsToRight;
        }

        return ans;
    }
};

File: 1769-minimum-number-of-operations-to-move-all-balls-to-each-box/1769-minimum-number-of-operations-to-move-all-balls-to-each-box.py
class Solution:
    def minOperations(self, boxes: str) -> List[int]:
        n = len(boxes)
        balls_to_left, moves_to_left = 0, 0
        balls_to_right, moves_to_right = 0, 0
        ans = [0] * n

        # Single pass: Calculate moves from both left and right
        for i in range(n):
            # Left pass
            ans[i] += moves_to_left
            balls_to_left += int(boxes[i])
            moves_to_left += balls_to_left

            # Right pass
            j = n - 1 - i
            ans[j] += moves_to_right
            balls_to_right += int(boxes[j])
            moves_to_right += balls_to_right

        return ans